### ISSUE TITLE
The problem of implementing Faster R-CNN

### ISSUE THREAD

**Issue by starimpact** [dmlc/mxnet#554]
Hi,
I found it's difficult directly to implement Faster R-CNN.
I think I must write two new kind of operators to handle the classification and regression problems.
Do I need to write them?

**Comment by winstywang** [comment]
I think currently the blocking one is ROI pooling. If you could do it, and contribute a PR, that is the best.

**Comment by starimpact** [comment]
The proposal progress is also needed to implement by writing two new operators, isn't?

**Comment by winstywang** [comment]
Yes, we definitely need them. However, they are relatively easier compare to ROI pooling,

**Comment by starimpact** [comment]
I think ROI pooling is easier, because it gradient-free.

**Comment by winstywang** [comment]
No, you need to back prop it as indicated in Fast RCNN paper.

**Comment by winstywang** [comment]
Anyway, we need to implement them all before we can get it. You contribution is always welcomed :)

**Comment by starimpact** [comment]
ROI pooling is actually the max pooling, definitely gradient-free(sure need backprop).
I think you didn't get my idea.:smile:

**Comment by piiswrong** [comment]
https://github.com/rbgirshick/py-faster-rcnn/tree/master/lib/roi_data_layer
Here is a python fast rnn implementation from Ross. Maybe we can copy some layers

**Comment by pluskid** [comment]
Related issue #214

**Comment by starimpact** [comment]
I want to implement the sibling layers as following, but it will die when I train it and without any error message:
Do you have any better ideas to figure it out?
```python 
    cls_hidlayer = MaxPoolingConv(last_cov, (1, 1), cls_node_num, 
                                  (1, 1), act_type='sigmoid', name='cls_hid')
    reg_hidlayer = MaxPoolingConv(last_cov, (1, 1), coord_node_num, 
                                  (1, 1), act_type='relu', name='reg_hid')

    ClsCost = AnchorClassifyCost(anchornum)
    myClsCost = ClsCost(data=cls_hidlayer, name = 'classify')
    RegCost = AnchorBBRegressionCost(anchornum, anchorbbs)
    myRegCost = RegCost(data=reg_hidlayer, name = 'regression')
    proposal_net = mx.sym.Group([myClsCost, myRegCost])
```

**Comment by starimpact** [comment]
I see the error when I use cpu:
```
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 314, in 'calling callback function'
  File "/home/mzhang/anaconda/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/operator.py", line 162, in backward_entry
    self.backward(in_data=tensors[0], out_data=tensors[1],
SystemError: Objects/cellobject.c:24: bad argument to internal function
```
what's wrong? :cry:

**Comment by starimpact** [comment]
```
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 314, in 'calling callback function'
  File "/home/mzhang/anaconda/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/operator.py", line 152, in forward_entry
    self.forward(in_data=tensors[0], out_data=tensors[1])
SystemError: Objects/cellobject.c:24: bad argument to internal function
```

**Comment by starimpact** [comment]
```python
    for ti in xrange(train_num):
        imgtmp = np.reshape(train_set_x[ti], (1, 1, stdsize[0], stdsize[1]))
        labeltmp = np.reshape(train_set_label[ti], 
                           (1, label_shape[0], label_shape[1], label_shape[2]))
        marktmp = np.reshape(train_set_mark[ti], 
                           (1, label_shape[0], label_shape[1], label_shape[2]))

        data[:] = mx.nd.array(imgtmp, ctx=ctxdev)
        label[:] = mx.nd.array(labeltmp, ctx=ctxdev)
        mark[:] = mx.nd.array(marktmp, ctx=ctxdev)

        proposal_exec.forward(is_train=True)
        #get cost
        outcost = np.mean(proposal_exec.outputs[0].asnumpy())
#        outcost = 0
        proposal_exec.backward()
        do_SGD(args, grads, lrn_now)
        cost_all += outcost
```

**Comment by starimpact** [comment]
it's my RPN, when I run the training of classification or regression individually, there is no problem, but not when combine them together. 
As you see, my classification cost layer and regression cost layer is implemented in python.

**Comment by starimpact** [comment]
I doubt whether something between python and mxnet has problem?

**Comment by starimpact** [comment]
I will write these two layers in c++ manner, check whether it is ok.

**Comment by vchuravy** [comment]
@starimpact You were trying to implement these layers in python? The error you a getting is certainly coming from the fact that something went bad in the NumpyOp layer. Can you share the code for your layers? (Best as a gist)

**Comment by starimpact** [comment]
```python

class AnchorClassifyCost(mx.operator.NumpyOp):
    """
    Classification On Each Anchor
    """
    def __init__(self, anchor_num):
        self.mininum = 1e-32
        self.anchornum = anchor_num
        super(AnchorClassifyCost, self).__init__(True)

    def list_arguments(self):
        """
        label is 0 or 1 multi-channel matrix, 1 means positive position
        """
        return ['data', 'label', 'mark_label']

    def list_outputs(self):
        return ['output']
#        return ['output', 'predict_output']

    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        label_shape = data_shape
        mark_shape = data_shape
        output_shape = data_shape
#        predict_shape = data_shape
        return [data_shape, label_shape, mark_shape], [output_shape]
#        return [data_shape, label_shape, mark_shape], [output_shape, predict_shape]

    def forward(self, in_data, out_data):
#        return
#        print 'classify foreward'
        datain = in_data[0]
        clstag = in_data[1]
        marktag = in_data[2]
        dataout = out_data[0]
#        predictout = out_data[1]
        costmatrix = -clstag * np.log(datain + self.mininum) - (1 - clstag) * np.log(1.0 - datain + self.mininum)
        costmatrix *= marktag
        np.copyto(dataout, costmatrix)
#        np.copyto(predictout, datain)

    def backward(self, out_grad, in_data, out_data, in_grad):
#        print 'len in_grad 0:', len(in_grad), in_grad[0].shape
#        return 
#        print 'classify backward'
        datain = in_data[0]
        clstag = in_data[1]
        marktag = in_data[2]
        gdata = in_grad[0]
        gvalues = -clstag / (datain + self.mininum) + (1 - clstag) / (1.0 - datain + self.mininum)
        gvalues *= marktag
        np.copyto(gdata, gvalues)
```

**Comment by starimpact** [comment]
i want to know how to debug when i encounter some questions.

**Comment by starimpact** [comment]
i also want to know how to make the error message always show up not just quit whith nothing.

**Comment by tqchen** [comment]
If you think this is the problem of numpy operator. Please try to simpilfy the script, for example
remove some of the operations performed.

It would be helpful write a complete short snippet that reproduces your problem. Both will help others to quickly understand what is going on your side.

**Comment by starimpact** [comment]
can you tell me what the error message means

**Comment by vchuravy** [comment]
It might be that the c-callback gets garbage-collected [1] That would explain why you don't see it all the time.

I am not familiar enough with python, but it might not keep a reference if the value is stored in a structure.

[1] http://stackoverflow.com/questions/24269939/systemerror-objects-cellobject-c24-bad-argument-to-internal-function

**Comment by starimpact** [comment]
@vchuravy  thank you anyway :smile:

**Comment by piiswrong** [comment]
@vchuravy this indeed could happen. 
@starimpact if this is your problem it should be fixed after the new PR is merged.

**Comment by vchuravy** [comment]
````
python -i numpy_softmax.py

>>> gc.get_referrers(mysoftmax.info_.forward)
[]
````

So if I am not mistaken that means that nothing is holding a reference to the elements in the `info_` structure and they might get gc'd. @starimpact I will create a PR and we can see if that solves your problem.

https://docs.python.org/2/library/gc.html
````
gc.get_referrers(*objs)
Return the list of objects that directly refer to any of objs. This function will only locate those containers which support garbage collection; extension types which do refer to other objects but do not support garbage collection will not be found.
````

**Comment by piiswrong** [comment]
@vchuravy I already pushed a PR

**Comment by starimpact** [comment]
I am implementing the cost layers in c++ manner, hope to solve it.

**Comment by starimpact** [comment]
but I don't know how to make a layer be a loss function.I know how to do with python, in which I can set ```__init__(False)```

**Comment by starimpact** [comment]
@vchuravy thank you, but your post from stachoverflow is too hard for me to understand.

**Comment by kaishijeng** [comment]
Starimpact,

   Are you able to run your faster-rcnn implementation properly? If yes, can you share your code?

Thanks,

**Comment by starimpact** [comment]
still one working.

**Comment by rodrigob** [comment]
I do not get why was this ticket closed. Was the feature implemented ?

**Comment by leefionglee** [comment]
@starimpact Hi, do you work it out?

**Comment by starimpact** [comment]
I finished the rpn part of faster-rcnn with cpp. Yes, I work it out and it works fine.

**Comment by vchuravy** [comment]
@starimpact Do you have the code online somewhere? Would be nice if we could incorporate it into mxnet

**Comment by starimpact** [comment]
In my github, there is a prototype code, but the execution ability is not guaranteed.

**Comment by pribadihcr** [comment]
@starimpact 

Could you give us the prototype code's link in your github?

thanks

**Comment by starimpact** [comment]
@pribadihcr https://github.com/starimpact/mxnet/tree/master/src/operator
please have a look the files which name prefixed with ```anchor```

**Comment by starimpact** [comment]
I have updated my mxnet, the rpn(files prefixed with anchor) laysers are what I'am using now.

**Comment by twtygqyy** [comment]
@starimpact Thanks for your implementation, could you provide some code about how to use the rpn layer?

**Comment by wangdelp** [comment]
@starimpact hi, have you submitted a PR request to MXNet. Do you have some examples to run the faster RCNN stuff? Thank you.

**Comment by starimpact** [comment]
@wangdelp It is just a part of Faster-RCNN. It is RPN.
I think it is a little complicated to use, so I am not sure whether it is suitable to PR.
